File: cpp-implementation/src/products/pcgames/Action.cpp
#include "Product.ih"

// Constructor implementation for action games
Action::Action(int id, string name, double price, int volume, string genre, 
               int ageRating, bool multiplayer, string supplier)
    : PC_Games(id, name, price, volume, genre, ageRating, supplier), 
      multiplayer(multiplayer) {}

// Calculates the discount depending on the current month and the packaged quantity
double Action::calculateDiscount(int volume, int month) const {
    if (month == 12) {
        return price * volume * 0.50;
    }
    return 0;
}

// Displays all the information 
string Action::displayProductInfo() const {
    ostringstream info;

    info << "ID: " << id << "\nnGame: " << name << "\nSupplier: " << getSupplier()
    << "\nGenre: " << genre << "\nMinimum Age: " << ageRating
    << "\nMultiplayer: " << (multiplayer ? "Yes" : "No") << "\nnPrice: €" << price
    << "\nVolume: " << volume << " units\n\n";

    return info.str();
}

File: java-implementation/src/main/java/com/inventory/products/pcgames/Action.java
package com.inventory.products.pcgames;

// Represents a concrete subclass of PC_Games
public class Action extends PC_Games {
    private boolean multiplayer;

    //  Constructs an instance with specified details with base class attributes included
    public Action(int id, String name, double price, int volume, String genre, int ageRating, boolean multiplayer, String supplier) {
        super(id, name, price, volume, genre, ageRating, supplier);
        this.multiplayer = multiplayer;
    }

    // Calculates the discount based on volume and month
    @Override
    public double calculateDiscount(int volume, int month) {
        if (month == 12) {
            return price * volume * 0.50;
        }
        return 0;
    }

    // Provides a string representation the details
    @Override
    public String displayProductInfo() {
        return "ID: " + id + "\nGame: " + name + "\nGenre: " + genre
               + "\nMinimum Age: " + ageRating + "\nMultiplayer ?: " + (multiplayer ? "Yes" : "No")
               + "\nPrice: €" + price + "\nVolume: " + volume + " units\nSupplier: " + supplier + "\n\n";
    }
}
